src/start/RealmInputScreen.js

```diff
@@ -7,36 +7,64 @@ import { useFocusEffect } from '@react-navigation/native';
 import type { RouteProp } from '../react-navigation';
 import type { AppNavigationProp } from '../nav/AppNavigator';
 import type { ApiResponseServerSettings } from '../api/settings/getServerSettings';
-import ErrorMsg from '../common/ErrorMsg';
 import ZulipTextIntl from '../common/ZulipTextIntl';
 import Screen from '../common/Screen';
 import ZulipButton from '../common/ZulipButton';
 import { tryParseUrl } from '../utils/url';
 import * as api from '../api';
 import { ThemeContext } from '../styles/theme';
 import { createStyleSheet, HALF_COLOR } from '../styles';
+import { showErrorAlert } from '../utils/info';
+import { TranslationContext } from '../boot/TranslationProvider';
+import type { LocalizableText } from '../types';
 
 type Props = $ReadOnly<{|
   navigation: AppNavigationProp<'realm-input'>,
   route: RouteProp<'realm-input', {| initial: boolean | void |}>,
 |}>;
 
-const urlFromInputValue = (realmInputValue: string): URL | void => {
+enum ValidationError {
+  InvalidUrl = 0,
+  NoUseEmail = 1,
+}
+function validationErrorMsg(validationError: ValidationError): LocalizableText {
+  switch (validationError) {
+    case ValidationError.InvalidUrl:
+      return 'Please enter a valid URL.';
+    case ValidationError.NoUseEmail:
+      return 'Please enter the server URL, not your email.';
+  }
+}
+
+type MaybeParsedInput =
+  | {| +valid: true, value: URL |}
+  | {| +valid: false, error: ValidationError |};
+const tryParseInput = (realmInputValue: string): MaybeParsedInput => {
   const withScheme = /^https?:\/\//.test(realmInputValue)
     ? realmInputValue
```

Comment:
nit: blank line between top-level declarations, here and above

I recognize that this type definition is closely related to the function. Still, this just feels visually harder to parse than it should be. I think it's that there's this mass of consecutive lines, so I expect the first of them to be the start of the function, and then it's not and instead the function starts several lines in.

(The exception to having such a blank line is when both declarations are a single line long. Perhaps also when "morally" a single line, like if there's a bunch of similar definitions in a row, most of them single lines, and a few of them run past 100 characters and therefore get Prettier-formatted to multiple lines.)